capellambse/decl.py

```diff
@@ -0,0 +1,404 @@
+    """Load an instruction stream from a YAML file.
+
+    Parameters
+    ----------
+    file
+        An open file-like object, or a path or PathLike pointing to such
+        a file. Files are expected to use UTF-8 encoding.
+    """
+    if hasattr(file, "read"):
+        file = t.cast(t.IO[str], file)
+        ctx: t.ContextManager[t.IO[str]] = contextlib.nullcontext(file)
+    else:
+        assert not isinstance(file, t.IO)
+        ctx = open(file, encoding="utf-8")
+
+    with ctx as opened_file:
+        return yaml.load(opened_file, Loader=YDMLoader)
+
+
+def apply(model: capellambse.MelodyModel, file: FileOrPath) -> None:
+    """Apply a declarative modelling file to the given model.
+
+    Parameters
+    ----------
+    model
+        The model to apply the instructions to.
+    file
+        An open file-like object to read YAML instructions from, or a
+        path to such a file. Files will be read with UTF-8 encoding.
+
+    Notes
+    -----
+    This function is not transactional: If an exception occurs during
+    this function, the model will be left partially modified, with no
+    reliable way to know how much of the YAML input has been consumed.
+    It is therefore advised to reload or discard the model immediately
+    in these cases, to avoid working with an inconsistent state.
+
+    Even though the YAML layout suggests linear execution from top to
+    bottom, the actual order in which modifications are executed is
+    implementation defined. This is necessary to support promises with
+    ``!promise``, but reorderings are still possible even if no promises
+    are used in an input document.
+    """
+    instructions = collections.deque(load(file))
+    promises = dict[Promise, capellambse.ModelObject]()
+    deferred = collections.defaultdict[Promise, list[_FutureAction]](list)
+
+    while instructions:
+        instruction = instructions.popleft()
+        parent = instruction.pop("parent")
+        if isinstance(parent, UUIDReference):
+            parent = model.by_uuid(parent.uuid)
+        for op_type, apply_op in _OPERATIONS.items():
+            try:
+                op = instruction.pop(op_type)
+            except KeyError:
+                continue
+            for promise, outcome in apply_op(promises, parent, op):
+                if isinstance(outcome, dict):
+                    deferred[promise].append(outcome)
+                else:
+                    if promise in promises:
+                        raise ValueError(
+                            f"promise_id defined twice: {promise.identifier}"
+                        )
+                    promises[promise] = outcome
+                    instructions.extend(deferred.pop(promise, ()))
+        if instruction:
+            keys = ", ".join(instruction)
+            raise ValueError(f"Unrecognized keys in instruction: {keys}")
+    if deferred:
+        raise UnfulfilledPromisesError(frozenset(deferred))
```

Comment:
This is very cool for me to read. I don't know if I shared that but I failed on implementing something with a Queue in the `apply_changeset.py`. `RuntimeError: deque mutated during iteration` >:(. I was very convinced that we'd need  a `Queue`-like structure since promises would be pushed back into it to get the proper solution. I suggest we take a closer look into special cases like re-locating `ModelObject`s and how this changes execution order of instructions. Explaining the effects of moving elements from top to bottom (in XML) and vice versa is a valuable insight to understand how the future `ChangeSet` technology and its dependency (this beautiful piece of code here) works together.

Reply:
> promises would be pushed back into it

Not quite. I've been there during development as well, and it ended in an infinite loop, because it would take the first instruction out, see that it can't resolve the promise, and put it back ... into an empty queue, so that the next instruction is the exact same one it just took out. :clown_face: So we need at least two queues, one to park promises that can't be resolved yet. (This is by the way exactly what `tests/test_decl.py::TestApplyPromises::test_unfulfilled_promises_raise_an_exception` is for.)

Reply:
> I suggest we take a closer look into special cases like re-locating `ModelObject`s and how this changes execution order of instructions.

It shouldn't affect the order of execution at all, because these are totally unrelated. But it could absolutely change the result of queries like `parent: la.root_function.functions[4]` (if those ever get implemented). We might even want to go as far as to resolve all `parent:` queries ahead of time, before modifying anything, in order to guarantee consistent results. But that's a discussion for another time :)

Reply:
> the effects of moving elements from top to bottom (in XML) and vice versa

The only effect that it should have (assuming no bugs) is for the moved element to show up somewhere else. This would of course affect all lists that would have found it and shift the indexes of elements after it.

I can see a potential use case for moving elements into a specific position in such a list, instead of plainly appending it at the end. Do you think that should be explored further, or would it not be necessary?

Reply:
As far as I'm aware, there was no interface spec for this use case yet. So I'd tend to agree with you and defer it to when someone (likely us) actually needs it, and we know exactly what the requirements are.